Storage in Image Management

nalind opened the pull request:
Picking up from https://github.com/kubernetes-incubator/cri-o/pull/28, this PR builds on #210 to pull images using containers/image, and to put images, pod sandboxes, and containers in space managed by containers/storage.

Close #298

runcom commented:
I'm wondering why Travis disappeared, @mrunalp do you know anything?

mrunalp commented:
Needs rebase.

nalind commented:
Ready.  Recently-landed changes meant that some of the functionality that was implemented in the storage cri branch was no longer needed, so it's been pruned out.

mrunalp commented:
There is a lint failure:
```
checking lint
cmd/server/main.go:88:3:warning: should use for range instead of for { select {} } (gosimple)
make: *** [Makefile:35: lint] Error 1
```

mrunalp commented:
From quick manual testing it works fine :) Now running tests locally and reviewing.

mrunalp commented on server/container.go:
```diff
@@ -341,13 +325,38 @@ func (s *Server) createSandboxContainer(containerID string, containerName string
+		return nil, err
+	}
+
+	mountPoint, err := s.storage.StartContainer(containerID)
```
What is the difference between Create and Start in storage? Creation of r/w layer?

nalind replied:
At the storage level, "Start" and "Stop" are more or less equivalent to "Mount" and "Umount".  The r/w layer is created in s.storage.CreatePodSandbox or s.storage.CreateContainer.  CreatePodSandbox also "Starts" (mounts) the infra container's layer.

nalind commented:
Rebased to master to resolve conflicts, fixed the lint error, and introduced a proposed `default_transport` configuration option which, while it's not really an accurate name, hopefully points people in the right direction: if an image name in a pull request can't be parsed as a pull source, we'll prepend the value and try again.

There's some additional logic to convert references to images in other transports into references suitable for use with the local storage when creating images and containers, so some of the self-tests still refer to images using the `docker://` prefix, and some don't, hopefully to better exercise the logic.

Reporting of image sizes is kind of broken - if you get an answer at all, it is probably wrong if any of the layers were compressed.

nalind commented on server/server.go:
```diff
@@ -233,15 +250,27 @@ func New(config *Config) (*Server, error) {
+		return nil, err
+	}
+
+	imageService, err := scri.GetStorageImageService(store, config.DefaultTransport)
```
We could.  I think it would actually be cleaner in terms of dependencies to move the entirety of `containers/storage/cri` into this repo, but I also want it to be in a place where other projects and reuse the logic instead of having to reimplement it themselves, and I'm not sure that's within the scope of what this repo wants to be.

Turning the image functions into members of an interface that isn't exactly `runtime.ImageServiceServer` would let us move some of its API glue here, and let `containers/storage` lose its hard dependency on `k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime`, which pulls in quite a few other things, so maybe that'd work better.

mrunalp replied:
> We could. I think it would actually be cleaner in terms of dependencies to move the entirety of containers/storage/cri into this repo, but I also want it to be in a place where other projects and reuse the logic instead of having to reimplement it themselves, and I'm not sure that's within the scope of what this repo wants to be.

I agree. IMO storage library shouldn't have any dependencies on the CRI API. That actually makes it harder to use from other projects. You could have convenience APIs to create a Rootfs for an image and a container ID but higher coupling than that will make it harder to reuse.

nalind replied:
Okay, just pushed changes which I think move us closer.  We still pull in kube by way of containers/image's support for openshift, but we're insulated from changes in the runtime API, at least.

mrunalp commented:
When I run make install on this PR, I see a bunch of SELinux related errors:
```
[root@dhcp-16-129 cri-o]# make install
install -D -m 755 ocid /usr/bin/ocid
/etc/selinux/targeted/contexts/files/file_contexts.bin:  line 1 error due to: Non-ASCII characters found
/etc/selinux/targeted/contexts/files/file_contexts.homedirs.bin:  line 1 error due to: Non-ASCII characters found
install -D -m 755 ocic /usr/bin/ocic
/etc/selinux/targeted/contexts/files/file_contexts.bin:  line 1 error due to: Non-ASCII characters found
/etc/selinux/targeted/contexts/files/file_contexts.homedirs.bin:  line 1 error due to: Non-ASCII characters found
install -D -m 755 conmon/conmon /usr/libexec/ocid/conmon
/etc/selinux/targeted/contexts/files/file_contexts.bin:  line 1 error due to: Non-ASCII characters found
/etc/selinux/targeted/contexts/files/file_contexts.homedirs.bin:  line 1 error due to: Non-ASCII characters found
install -D -m 755 pause/pause /usr/libexec/ocid/pause
/etc/selinux/targeted/contexts/files/file_contexts.bin:  line 1 error due to: Non-ASCII characters found
/etc/selinux/targeted/contexts/files/file_contexts.homedirs.bin:  line 1 error due to: Non-ASCII characters found
install -d -m 755 /usr/share/man/man{8,5}
install -m 644 docs/ocid.8 -t /usr/share/man/man8
/etc/selinux/targeted/contexts/files/file_contexts.bin:  line 1 error due to: Non-ASCII characters found
/etc/selinux/targeted/contexts/files/file_contexts.homedirs.bin:  line 1 error due to: Non-ASCII characters found
install -m 644 docs/ocid.conf.5 -t /usr/share/man/man5
/etc/selinux/targeted/contexts/files/file_contexts.bin:  line 1 error due to: Non-ASCII characters found
/etc/selinux/targeted/contexts/files/file_contexts.homedirs.bin:  line 1 error due to: Non-ASCII characters found
install -D -m 644 ocid.conf /etc/ocid.conf
/etc/selinux/targeted/contexts/files/file_contexts.bin:  line 1 error due to: Non-ASCII characters found
/etc/selinux/targeted/contexts/files/file_contexts.homedirs.bin:  line 1 error due to: Non-ASCII characters found
```
cc: @rhatdan

mrunalp commented:
We have a lot of lint errors in container/storage vendoring.

rhatdan commented:
Looks like you got garbage in 
/etc/selinux/targeted/contexts/files/file_contexts.bin:  line 1 error due to: Non-ASCII characters found
/etc/selinux/targeted/contexts/files/file_contexts.homedirs.bin:  line 1 error due to: Non-ASCII characters found

Just remove these files, they should get recreated.

mrunalp commented:
@rhatdan That fixed it for me. Thanks!

nalind commented:
@mrunalp Yeah, it looks like the linters are having trouble with cgo code.

mrunalp commented:
@runcom Do you have any thoughts on how to deal with the lint issues. Are they fixable by modifying the code.

runcom commented:
Can we skip with grep -v the whole vendored code?

mrunalp commented:
@runcom @nalind Yeah, makes sense to skip vendored code in this repo. We can fixup what we can in containers/storage separately.

runcom commented:
@mrunalp @nalind are we holding this one a little more so Nalin could figure out how to remove the containers/image dependency or can I go ahead and review it?

mrunalp commented:
@runcom You can wait for the removal of containers/image dependency.

runcom commented on server/image_status.go:
```diff
@@ -9,8 +9,20 @@ import (
+	if img != nil {
+		image = img.GetImage()
+	}
+	status, err := s.images.ImageStatus(ctx, image)
```
something must be wrong in here since with k8s I'm getting:
```
 Error syncing pod, skipping: failed to "StartContainer" for "nginx" with ImageInspectError: "Failed to inspect image \"nginx\": rpc error: code = 2 desc = image not known"
```

mrunalp replied:
@runcom Yeah, I also ran into this issue while testing with k8s with this PR this week. I think we should maintain a wiki page for k8s integration setup and issues.

runcom replied:
@mrunalp :+1: let's do that once we have this PR merged and fix issue by issue the bug we found while integrating with k8s

nalind replied:
Was the image already pulled when you encountered this error?  If not, if you manually pull it, does that make the error stop occurring?

runcom replied:
@nalind I believe k8s itself would have already pulled it. I'd say let's not worry about this issue _right now_ an defer it to when this PR is fully ready to be reviewed and integrated. As Mrunal pointed out there may be other issues, but would be simple to tackle each if we have this PR in a good shape to work on when merged.

runcom replied:
found out, we basically need to check if the `err == storage.ErrImageUnknown` and `return &pb.ImageStatusResponse{}` (the empty response) in order to have k8s calling `pull_image`. The CRI api says to return `nil` but that's probably wrong I guess :confused: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/api/v1alpha1/runtime/api.proto#L73

runcom replied:
fixing CRI api comment here https://github.com/kubernetes/kubernetes/pull/38589

runcom commented:
@nalind something should be wrong in the cleanup stuff, basically if I "ctrl-c" ocid, I'm left with:
```
/var/lib/ocid/devicemapper
/var/lib/ocid/devicemapper/mnt/798a44ab34fa878b2d3ff9934cf61a0a235ebf4ea3f19dd933caa6b60e987f29
/var/lib/ocid/devicemapper/mnt/7bc934fa2568dc5504c216876b47a8abd73ded0d7aa5bfb2127f12e38a1a6d59
```
mounted

runcom commented on server/sandbox_run.go:
```diff
@@ -20,6 +20,7 @@ import (
 
 // RunPodSandbox creates and runs a pod-level sandbox.
 func (s *Server) RunPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest) (*pb.RunPodSandboxResponse, error) {
+	s.Update()
```
It's pretty awful to have this check on every request but I can see it's needed. BTW, should this check the error and log it at least?

nalind replied:
Yeah, the logging should be using Error priority, not Warning as it is now.  Changing it.

runcom commented on cmd/server/config.go:
```diff
@@ -22,17 +24,21 @@ var commentedConfigTemplate = template.Must(template.New("config").Parse(`
-sandbox_dir = "{{ .SandboxDir }}"
+# run is a path to the "run directory". OCID stores all of its state
+# in this directory.
+runroot = "{{ .RunRoot }}"
```
do we really want to customize things under `/var/run/...`? I would not

runcom commented on cmd/server/config.go:
```diff
@@ -11,9 +10,12 @@ import (
-	pausePath           = "/usr/libexec/ocid/pause"
+	pauseImage          = "kubernetes/pause"
+	pauseCommand        = "/pause"
+	defaultTransport    = "docker://"
```
this would be implied at some point by the image reference coming from k8s I believe, I'm not sure why we need this in the config

runcom commented on cmd/server/config.go:
```diff
@@ -11,9 +10,12 @@ import (
+	ocidRunRoot         = "/var/run/containers"
 	conmonPath          = "/usr/libexec/ocid/conmon"
-	pausePath           = "/usr/libexec/ocid/pause"
+	pauseImage          = "kubernetes/pause"
```
I believe this was an issue with @cyphar as there's the need to control network access and having a pause path with a statically linked binary was wanted

cyphar replied:
Yeah, so my main concern is that I wanted to make sure that it is possible to use `ocid` **without** needing to pull images from anywhere. This is something that will be a requirement from SUSE's side eventually (so we can audit all of the base images).

runcom replied:
right, I believe we should have still a `pausePath` flag for now in order to setup pause directly w/o pulling anything from the net. @cyphar does that sound good? we'll have either the option to pull the image or run with whatever `pausePath` an admin provides

cyphar replied:
:+1: We can revisit this later (after this PR hits). In particular, a proper full-on solution would involve being able to alias images to a local directory which has OCI layouts in it -- so you could still specify `docker://abc` and that would map to some image already on the filesystem.

runcom commented on server/container_create.go:
```diff
@@ -110,17 +100,16 @@ func (s *Server) CreateContainer(ctx context.Context, req *pb.CreateContainerReq
 		return nil, errors.New("createSandboxContainer needs a sandbox")
 	}
+
+	// TODO: factor generating/updating the spec into something other projects can vendor
```
`specgen` it's actually already a vendored package :confused: (maybe we need to cleanup this function instead)

mrunalp replied:
This comment can be removed as we are already using runtime-tools generate library. As @runcom pointed out we could instead move creation of spec to another helper function.

runcom commented on server/image_list.go:
```diff
@@ -9,8 +9,25 @@ import (
+		response.Images = append(response.Images, &pb.Image{
+			Id:       sPtr(result.ID),
+			RepoTags: result.Names,
+			Size_:    result.Size,
```
is `Size_` with the underscore something from `grpc`?!

nalind replied:
Yes.

runcom commented on server/image_pull.go:
```diff
@@ -1,82 +1,38 @@
-		return nil, err
+	// TODO(runcom?): deal with AuthConfig in req.GetAuth()
+	// TODO(somebody?): either rework PullImage to verify signatures, or do it by using PullImageUsingContexts, if that's enough
+	// TODO: what else do we need here? (Signatures when the story isn't just pulling from docker://)
```
I'd leave signatures off from this first iteration...

runcom commented on server/image_pull.go:
```diff
@@ -1,82 +1,38 @@
+		image = img.GetImage()
 	}
-	dir, err := directory.NewReference(filepath.Join(s.config.ImageDir, tr.StringWithinTransport()))
+	systemContext := &types.SystemContext{}
```
There's a field in `types.SystemContext` called `DockerAuthConfig` which should be applied when pulling from `docker://` (this should take care of https://github.com/kubernetes-incubator/cri-o/pull/189/files#diff-a2128ff22d4b287bcce94b33ec7c0084R15)

runcom commented on server/image_pull.go:
```diff
@@ -1,82 +1,38 @@
+	// TODO(runcom?): deal with AuthConfig in req.GetAuth()
+	// TODO(somebody?): either rework PullImage to verify signatures, or do it by using PullImageUsingContexts, if that's enough
+	// TODO: what else do we need here? (Signatures when the story isn't just pulling from docker://)
+	image := ""
```
should we check `image == nil` and error out if that's the case? instead of leaving to `RemoveImage`

runcom commented on server/image_status.go:
```diff
@@ -9,8 +9,20 @@ import (
-	// containers/storage will take care of this by looking inside /var/lib/ocid/images
-	// and getting the image status
-	return &pb.ImageStatusResponse{}, nil
+	image := ""
```
should we check `image == nil` and error out if that's the case? instead of leaving to `RemoveImage`

runcom commented on server/sandbox_run.go:
```diff
@@ -35,6 +37,10 @@ func (s *Server) RunPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
 	if err != nil {
 		return nil, err
 	}
+	_, containerName, err := s.generateContainerIDandName(name, "infra", 0)
```
the last `0` should be from `GetAttempt` from the request? /cc @mrunalp

nalind replied:
Okay, retrieving it.

runcom commented on server/server.go:
```diff
@@ -46,7 +50,7 @@ type Server struct {
 
 func (s *Server) loadContainer(id string) error {
-	config, err := ioutil.ReadFile(filepath.Join(s.runtime.ContainerDir(), id, "config.json"))
+	config, err := s.store.GetContainerDirectoryFile(id, "config.json")
```
maybe `GetFromContainerDirectory`?

runcom commented on server/container_create.go:
```diff
@@ -110,17 +100,16 @@ func (s *Server) CreateContainer(ctx context.Context, req *pb.CreateContainerReq
-
 	args := containerConfig.GetArgs()
 	if args == nil {
 		args = []string{"/bin/sh"}
```
@nalind you should fix this by setting the args coming from the image's config if the `containerConfig.GetArgs()` is `nil`. Otherwise, unless you run `kubectl run nginx --image=nginx -- <arg1> <arg2> ... <argN>` you just get a shell instead of running nginx.

nalind replied:
Okay, assuming we get back a valid `github.com/opencontainers/image-spec/specs-go/v1.Image` in the image's config blob, we can do that.

jjlakis commented on server/server.go:
```diff
@@ -168,31 +178,136 @@ func (s *Server) loadSandbox(id string) error {
+	})
+	for deadPodContainer := range deadPodContainers {
+		// forget this container
+		c := s.getContainer(deadPodContainer)
```
Not sure if it's about some files I left on my machine, but I had to add `<nil>` check here to prevent panic on the next line, while trying to access reference that not exists. What conditions has to be met to treat container as dead?

nalind replied:
"dead" is probably the wrong word - it's the list of containers that we knew about which are no longer in the list that's kept on disk, presumably because they were shut down or died and were removed by some other party.

nalind replied:
Okay, I think I've got this one - when removing a pod sandbox and its containers, we weren't removing the container or pod sandbox IDs from their respective ID indices.

feiskyer commented:
Tried running this patch with kubelet, a simple nginx pod start failed (kubelet log):

```
I1213 10:38:05.982481   18019 server.go:674] Event(v1.ObjectReference{Kind:"Pod", Namespace:"default", Name:"nginx-272207814-gmg98", UID:"3b230c4e-c120-11e6-a418-fa163e66591a", APIVersion:"v1", ResourceVersion:"362", FieldPath:""}): type: 'Normal' reason: 'SandboxReceived' Pod sandbox received, it will be created.
E1213 10:38:10.348822   18019 remote_runtime.go:78] RunPodSandbox from runtime service failed: rpc error: code = 2 desc = error creating pod sandbox with name "default-nginx-272207814-gmg98-0": Error writing blob: ApplyLayer wait: no child processes stdout: {"layerSize":239840}
 stderr:
E1213 10:38:10.348891   18019 kuberuntime_sandbox.go:54] CreatePodSandbox for pod "nginx-272207814-gmg98_default(3b230c4e-c120-11e6-a418-fa163e66591a)" failed: rpc error: code = 2 desc = error creating pod sandbox with name "default-nginx-272207814-gmg98-0": Error writing blob: ApplyLayer wait: no child processes stdout: {"layerSize":239840}
 stderr:
E1213 10:38:10.348911   18019 kuberuntime_manager.go:643] createPodSandbox for pod "nginx-272207814-gmg98_default(3b230c4e-c120-11e6-a418-fa163e66591a)" failed: rpc error: code = 2 desc = error creating pod sandbox with name "default-nginx-272207814-gmg98-0": Error writing blob: ApplyLayer wait: no child processes stdout: {"layerSize":239840}
 stderr:
```

And there is nothing help in ocid logs.

@nalind Do you know what's the possible cause?

runcom commented:
> @nalind Do you know what's the possible cause?

@feiskyer please follow my instruction in https://gist.github.com/runcom/ba58bf2f64e38d9f5d376d587751a0f9 (for Centos 7) on how to test this pull request.

nalind commented on server/image_pull.go:
```diff
@@ -1,82 +1,38 @@
-	// save manifest
-	m, _, err := i.Manifest()
+	options := &copy.Options{}
+	_, err = s.images.PullImageUsingContexts(image, policyContext, options)
```
The updated PullImage() will take a SystemContext, dig the default policy out of it, and use a new context based off of that.

runcom commented:
something is going wrong when ocid is restarted and we request to pull an image we already pulled before the restart:
```
13:03:16 [github.com/kubernetes-incubator/cri-o] ‹demo› sudo ./ocic image list
ID: c3aa2798e0100e1f0090cabd56aff8476e7e0bd2d211452e6c1dabd86cbbddaf
Tag: docker.io/kubernetes/pause:latest
Tag: kubernetes/pause
Tag: docker.io/kubernetes/pause
ID: c7c264a871489ed302de80067fb561e8b88c90d3acf910b234bf643bb09a3655
ID: 22bae1e7997fdddd6e9a2ad7d8c771eb9bef76cfd5cdebb8e52d80443ab001b4
Tag: docker.io/library/redis:latest
Tag: redis:latest
Tag: docker.io/library/redis
ID: 150c7a727048577299507032145f9e4bee22135c69abb9dc2172276a6324d42f
Tag: docker.io/library/nginx:latest
Tag: nginx:latest
Tag: docker.io/library/nginx
ID: 6cee24dae6d1292d905f07a293f811affa79287e0820d65b8d2217aa06024d2c
Tag: docker.io/library/redis:latest
Tag: redis:latest
Tag: docker.io/library/redis
``

runcom commented:
BTW, my https://github.com/runcom/ocid-1/tree/demo branch is in good shape building from this PR :+1: (stopping ocid and restarting it now seems to work also and containers get re-synced with kubelet also!)

nalind commented:
Updated to vendor containers/image's master branch.

runcom commented:
Tests hang on my system with this PR (while I can run tests just fine with #210)

nalind commented:
@runcom I'm not _entirely_ sure it's all about the pulling of images (I think, but haven't verified, that the integration tests were faster in this branch as recently as a few days ago), but I'm working on a test helper which will import the locally-built "pause" binary into storage as an image.

nalind commented:
Using a bin2img helper seems to be helping.  I suppose it also makes the code easier to test on other architectures, which is a bonus I hadn't thought about.

runcom commented:
@nalind this could be rebased/reworked based on #210 now

nalind commented:
Okay, rebased.  The updated version of `containers/image` needed to have `pkg/error` added to the dependencies list.

runcom commented:
Great thanks @nalind! I'll test this our by the end of the day

mrunalp commented:
@nalind Looks like we can merge the first 2 commits separately as they are fixing issues outside of the integration. Could you create separate PRs for those?

nalind commented:
Will do.

nalind commented:
Rebase is ready.  Thanks for merging #291 so quickly.

runcom commented:
Tests are still super slow for me :cry: not sure it's something with devicemapper since I see a bunch of:
```
Dec 20 16:34:48 runcom.usersys.redhat.com kernel: XFS (dm-6): Mounting V5 Filesystem
Dec 20 16:34:48 runcom.usersys.redhat.com kernel: XFS (dm-6): Ending clean mount
Dec 20 16:35:32 runcom.usersys.redhat.com kernel: XFS (dm-6): Unmounting Filesystem
Dec 20 16:35:35 runcom.usersys.redhat.com kernel: XFS (dm-6): Mounting V5 Filesystem
Dec 20 16:35:35 runcom.usersys.redhat.com kernel: XFS (dm-6): Ending clean mount
Dec 20 16:35:35 runcom.usersys.redhat.com kernel: XFS (dm-6): Unmounting Filesystem
Dec 20 16:35:37 runcom.usersys.redhat.com kernel: XFS (dm-6): Mounting V5 Filesystem
Dec 20 16:35:37 runcom.usersys.redhat.com kernel: XFS (dm-6): Ending clean mount
```
But just one test takes around 1m. @nalind which graphdriver are you using?

nalind commented:
By default, I'm also using device mapper.  The localintegration suite finishes in around 5 to 7 minutes in my VM.

runcom commented:
> By default, I'm also using device mapper. The localintegration suite finishes in around 5 to 7 minutes in my VM.

I test on my host and here is the time it takes for me (laptop is pretty fast):
```
16:32:54 [github.com/kubernetes-incubator/cri-o] ‹pr/origin/189› sudo make localintegration
go build --tags "selinux seccomp  " -o ocid ./cmd/server/
go build -o ocic ./cmd/client/
go build -o kpod ./cmd/kpod/
make -C conmon
make[1]: Entering directory '/home/amurdaca/go/src/github.com/kubernetes-incubator/cri-o/conmon'
make[1]: 'conmon' is up to date.
make[1]: Leaving directory '/home/amurdaca/go/src/github.com/kubernetes-incubator/cri-o/conmon'
make -C pause
make[1]: Entering directory '/home/amurdaca/go/src/github.com/kubernetes-incubator/cri-o/pause'
make[1]: 'pause' is up to date.
make[1]: Leaving directory '/home/amurdaca/go/src/github.com/kubernetes-incubator/cri-o/pause'
make -C test/bin2img
make[1]: Entering directory '/home/amurdaca/go/src/github.com/kubernetes-incubator/cri-o/test/bin2img'
go build -o bin2img
make[1]: Leaving directory '/home/amurdaca/go/src/github.com/kubernetes-incubator/cri-o/test/bin2img'
./test/test_runner.sh
++ time bats --tap .
1..39
ok 1 # skip (skip this test since apparmor is not enabled.) load default apparmor profile and run a container with it
ok 2 # skip (skip this test since apparmor is not enabled.) load a specific apparmor profile as default apparmor and run a container with it
ok 3 # skip (skip this test since apparmor is not enabled.) load default apparmor profile and run a container with another apparmor profile
ok 4 # skip (skip this test since apparmor is not enabled.) run a container with wrong apparmor profile name
ok 5 # skip (skip this test since apparmor is not enabled.) run a container after unloading default apparmor profile
ok 6 ctr remove
ok 7 ctr lifecycle
ok 8 ctrs status for a pod
ok 9 ctr list filtering
ok 10 ctr list label filtering
ok 11 ctr metadata in list & status
ok 12 ctr execsync
ok 13 ctr execsync failure
ok 14 ctr stop idempotent
ok 15 image pull
ok 16 image list with filter
ok 17 image list/remove
ok 18 image status/remove
ok 19 pod release name on remove
ok 20 pod remove
ok 21 pod list filtering
ok 22 pod metadata in list & status
ok 23 pass pod sysctls to runtime
ok 24 pod stop idempotent
ok 25 pod remove idempotent
ok 26 pod stop idempotent with ctrs already stopped
ok 27 restart ocid and still get pod status
ok 28 ocid restore
ok 29 ocic runtimeversion
ok 30 ctr seccomp profiles unconfined
ok 31 ctr seccomp profiles runtime/default
ok 32 ctr seccomp profiles wrong profile name
ok 33 # skip (need https://issues.k8s.io/36997) ctr seccomp profiles localhost/profile_name
ok 34 ctr seccomp profiles falls back to pod profile
ok 35 ctr seccomp profiles falls back to unconfined
ok 36 pod seccomp profiles unconfined
ok 37 pod seccomp profiles runtime/default
ok 38 pod seccomp profiles wrong profile name
ok 39 # skip (need https://issues.k8s.io/36997) pod seccomp profiles localhost/profile_name

real    28m52.778s
user    1m37.085s
sys     0m49.643s
```

cyphar commented:
On the topic of integration tests, the current system of integration testing leaves an **incredible** amount of crap on my host. You've got `nsfs` bindmounts that haven't been cleaned up, `runc` containers that clash with the test configurations, `shm` mounts that cause further retests to fail, etc.

We really need to fix this, because it's making it hard for people (at least for me) to actually test that their code works...

runcom commented:
> On the topic of integration tests, the current system of integration testing leaves an incredible amount of crap on my host. You've got nsfs bindmounts that haven't been cleaned up, runc containers that clash with the test configurations, shm mounts that cause further retests to fail, etc.
> 
> We really need to fix this, because it's making it hard for people (at least for me) to actually test that their code works...

you're right @cyphar - after this is merged (hopefully) we can fully run the integration tests in a docker container (with the Dockerfile). Hopefully that helps, or not?

mrunalp commented:
I am seeing the same slow tests that @runcom reported. Maybe we should cache the images for the duration of the test suite. 30 min tests aren't going to help with development :/

mrunalp commented:
How about dropping image pull from container create? That way we follow the k8s model of first having to pull the image? (We can add enhancements to ocic to pull first if required).

mrunalp commented on cmd/server/config.go:
```diff
@@ -11,9 +10,12 @@ import (
 
 const (
-	ocidRoot            = "/var/lib/ocid"
+	ocidRoot            = "/var/lib/containers"
```
Should we use `/var/lib/ocid` here?

rhatdan replied:
I would prefer to leave this containers, since we want the storage to be shared by other container runtimes.

mrunalp commented on cmd/server/config.go:
```diff
@@ -11,9 +10,12 @@ import (
 const (
-	ocidRoot            = "/var/lib/ocid"
+	ocidRoot            = "/var/lib/containers"
+	ocidRunRoot         = "/var/run/containers"
```
Should we use `/var/run/ocid` here?

rhatdan replied:
/run/ocid would be correct here.

mrunalp commented on cmd/server/config.go:
```diff
@@ -123,8 +135,9 @@ func DefaultConfig() *server.Config {
-			ImageDir: filepath.Join(ocidRoot, "store"),
+			DefaultTransport: defaultTransport,
+			PauseImage:       pauseImage,
+			PauseCommand:     pauseCommand,
```
Can't we get the command from the image itself?

mrunalp replied:
We could handle this later, though.

nalind replied:
If there's a config blob and it can be decoded as an image-spec v1 image, yes.  If that's not every image, we'll probably have to have some default.

mrunalp commented on server/config.go:
```diff
@@ -23,17 +23,20 @@ type Config struct {
-	// state and other information.
-	ContainerDir string `toml:"container_dir"`
+	// StorageOption is a list of storage driver specific options.
+	StorageOption []string `toml:"storage_option"`
```
nit: StorageOption -> StorageOptions

nalind replied:
Changing it.

mrunalp commented on server/sandbox_run.go:
```diff
@@ -306,23 +322,21 @@ func (s *Server) RunPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
 
-	err = g.SaveToFile(filepath.Join(podSandboxDir, "config.json"), generate.ExportOptions{})
+	saveOptions := generate.ExportOptions{}
+	mountPoint, err := s.storage.StartContainer(id)
```
Are we missing a defer to remove this if any subsequent calls in the function fail?

nalind replied:
The directory where we're storing the config.json file should get removed along with rest of the pod sandbox storage whenever it gets deleted, so I think this is fine.

mrunalp commented on server/container_create.go:
```diff
@@ -62,27 +61,17 @@ func (s *Server) CreateContainer(ctx context.Context, req *pb.CreateContainerReq
 		if err != nil {
 			s.releaseContainerName(containerName)
-			err1 := os.RemoveAll(containerDir)
+			err1 := s.storage.DeleteContainer(containerID)
```
This should probably be setup in a defer after createSandboxContainer otherwise there is no guarantee whether storage was even setup for the container.

mrunalp commented on server/container_start.go:
```diff
@@ -2,21 +2,49 @@ package server
+	if err = specgen.SaveToFile(filepath.Join(workdir, "config.json"), saveOptions); err != nil {
+		return nil, fmt.Errorf("failed to rewrite template configuration for container: %v", err)
+	}
+	if err = specgen.SaveToFile(filepath.Join(rundir, "config.json"), saveOptions); err != nil {
```
I don't quite get what we are trying to achieve here. At this point the container is already created so from runc's perspective a process has already been created so changing config.json won't have any effect.

nalind replied:
At this point we've mounted the rootfs (which theoretically is allowed to change from use to use), so we know where it is.  Here we update the configuration to reflect the current location, in case it changed since the last time we started the container.  We don't spawn runc until the later call to runtime.StartContainer.

nalind commented:
@mrunalp This revision includes some of the changes discussed above, but my attempts to work around requiring each test run to pull down its own copy of the redis image, if it used it, turned up some sync problems in the storage library's devicemapper driver, so it still implicitly pulls images that it doesn't have when it is asked to create a container.

While those get tended to, the integration tests run to completion for me when I adjust `helpers.bash` to either set `$STORAGE_DRIVER` to a different driver name, or have it invoke both `ocid` and `bin2img` with their `--storage-driver` options (the options we add to ocid now use `-` instead of `_` in their names, to match the style of other option names that are already provided by ocid) to have the same effect.

mrunalp commented:
@nalind I am giving it a try. Thanks!

mrunalp commented:
The tests passed for me but took 23 minutes. @nalind Do you think overlay will work if we run all tests on a single ocid instance without teardown for each test? We can try to make that change and rename a pod per test so there are no clashes and then we could clobber everything at the end of all the tests.

cc: @runcom

runcom commented:
> The tests passed for me but took 23 minutes. @nalind Do you think overlay will work if we run all tests on a single ocid instance without teardown for each test? We can try to make that change and rename a pod per test so there are no clashes and then we could clobber everything at the end of all the tests.

I think we could have a single ocid running, each test makes sure to cleanup pods and containers without removing the ocid root. This would allow us to keep stored images across tests I believe.

mrunalp commented:
@runcom Could you run the tests on this PR with overlay driver?

runcom commented:
> @runcom Could you run the tests on this PR with overlay driver?

yup, testing right now

runcom commented:
@nalind could you add `test/bin2img/bin2img` to the `.gitignore` also?

runcom commented:
@mrunalp tests run fine, the issue is that they're still slow even with overlay (and overlay2)

mrunalp commented:
@runcom Yeah, let's figure out how to fix that in bats my not tearing down per test.

rhatdan commented:
@runcom We will wait to fix those issues until after break, so we will not merge, but we can carry a patch so people can start experimenting with a packaged version of OCID on Rawhide.
For example I can make sure it works with SELinux in enforcing mode and fix container-selinux to make it happen.  We also need to start working on docker-storage-setup and convirt it to container-storage-setup so it will work with both container runtimes.

runcom commented:
> @runcom We will wait to fix those issues until after break, so we will not merge, but we can carry a patch so people can start experimenting with a packaged version of OCID on Rawhide.
> For example I can make sure it works with SELinux in enforcing mode and fix container-selinux to make it happen. We also need to start working on docker-storage-setup and convirt it to container-storage-setup so it will work with both container runtimes.

SGTM

runcom commented:
@nalind can you cherry-pick here the top most 3 commits from https://github.com/runcom/ocid-1/commits/alpha? So this PR works with basic k8s pod functionalities

runcom commented:
@nalind when starting ocid with overlay2 and docker started already with overlay2 I see:
```
$ mount
[...]
/dev/mapper/fedora-root on /var/lib/docker/overlay2 type ext4 (rw,relatime,seclabel,data=ordered)
/dev/mapper/fedora-root on /var/lib/ocid/overlay2 type ext4 (rw,relatime,seclabel,data=ordered)
```
is this expected?

surajssd commented:
@mrunalp 
Direct pulling this PR and trying to build it caused this error:

```bash
✔ ~/go/src/github.com/kubernetes-incubator/cri-o [storage|…9] 
07:18 $ make
mkdir -p /home/vagrant/go/src/github.com/kubernetes-incubator/cri-o/vendor/src/github.com/kubernetes-incubator
ln -sfn /home/vagrant/go/src/github.com/kubernetes-incubator/cri-o /home/vagrant/go/src/github.com/kubernetes-incubator/cri-o/vendor/src/github.com/kubernetes-incubator/cri-o
go build --tags "selinux seccomp  " -o ocid ./cmd/server/
# github.com/mtrmac/gpgme
vendor/src/github.com/mtrmac/gpgme/data.go:4:20: fatal error: gpgme.h: No such file or directory
 // #include <gpgme.h>
                    ^
compilation terminated.
Makefile:58: recipe for target 'ocid' failed
make: *** [ocid] Error 2
```

So installed package: `gpgme-devel` and `libassuan-devel` and then build successful.
But this did not fix https://github.com/kubernetes-incubator/cri-o/issues/299 either.

nalind commented:
@runcom okay, I'm assuming that you also want the ones that you added after https://github.com/kubernetes-incubator/cri-o/pull/189#issuecomment-268641503.  The graph drivers would make whichever location they're using as their top level private, so if you're using both docker and ocid, what you're seeing in https://github.com/kubernetes-incubator/cri-o/pull/189#issuecomment-268642176 makes sense.

nalind commented:
@surajssd When you say this doesn't fix #299, which method are you using to list images?

nalind commented:
To remove the dependency on having a `/etc/containers/policy.json` in place that lets us run the tests, this revision adds a `--signature-policy` switch to ocid, with comments in the default configuration file which encourage it to not be set, and uses the one from the `test` directory when running tests.  The default config also used `-` instead of `_` for some of the settings which this PR was adding, which didn't match the tags that indicated which settings it needed to read in, but I think I've straightened that out.

mtrmac commented on test/bin2img/bin2img.go:
```diff
@@ -0,0 +1,194 @@
+
+	storage.Transport.SetStore(store)
+	imgName := storage.Transport.Name() + ":" + imageName
+	ref, err := transports.ParseImageName(imgName)
```
(Random drive-by comment): This can skip a step and do
```go
ref, err := storage.Transport.ParseReference(imageName)
```
or perhaps even
```go
ref, err := storage.Transport.ParseStoreReference(store, imageName)
```
(not sure whether there is a difference between the two).

No need to splice the strings just to split them again.

nalind replied:
Right, we can just use `ParseStoreReference` instead of a `SetStore`/`ParseImageName` combination.  Changing it.

mtrmac commented on test/bin2img/bin2img.go:
```diff
@@ -0,0 +1,194 @@
+	if err != nil {
+		logrus.Errorf("error preparing to write image: %v", err)
+		os.Exit(1)
+	}
```
Pedantically, `defer img.Close()` (or perhaps even a version with error checking) should happen here; at least right now `storageImageDestination.Close` is empty so this makes no difference.

nalind replied:
Makes sense.  Changing it.

nalind commented:
Updated with another stab at the underscores/hyphens correction, and to remove redundant calls to the storage StartContainer (mount) and StopContainer (unmount) methods, neither of which should be attempted when the runtime monitor is running and using the container's rootfs.  We mount a container's rootfs after creating it, and unmount it right before destroying it.

nalind commented:
Updated the README per conversation in #298.

nalind commented:
Per out-of-band conversation, I've added a `copyimg` tool which we use to pull down the `redis` image at the start of a test run (if it's been removed by `make clean`), stash it in `test/testdata/redis-image`, and import it from there before starting `ocid`.  It and the `bin2img` tool now use the `github.com/urfave/cli` package for parsing command line options to be more in line with `ocid` itself.  I'm hoping this'll speed up integration test runs.

Since we no longer depend on the implicit pulling of images, and I'm told that's not expected by CRI, that logic is no longer there.  I'm not entirely sure I like that, though, since it only adds a repetitive step for people who are trying to drive it with `ocic`.

Tests for image pulling now pass a flag to the `start_ocid` function to tell it to not build an image for `pause`, so that if `ocic image pull` fails, it's more obvious to the test.

Lastly, I added a check that I forgot to add earlier, skipping the updating of the pod and container lists when the storage library tells us that the set of Containers it keeps hasn't been changed since the last time we changed it ourselves or checked for changes made by others.  This should save some time.

mrunalp commented:
okay, tests run ~4 minutes now :+1: I will review the new bits. 

@runcom Can you also review/test this?

runcom commented:
Alright, tests run in reasonable time now :+1: Thanks @nalind 

Since the other commits are mine, I'm gonna just test the integration with k8s before acking on this PR for the last time hopefully :) @jjlakis tried this PR and he was successful when running with k8s though :)

runcom commented:
storage is busted for me :( I cannot run with `overlay2`:
```
DEBU[2017-01-12 17:08:36.249872332+01:00] [graphdriver] trying provided driver "overlay2"
```
`ocid` is stuck at that in the log file and doesn't proceed. I've cleaned up everything so the state on my laptop should be good (tried more than twice...)

Plus, when testing this PR with devicemapper AND k8s there's some kind of deadlock when k8s calls `ListImage`:
```
DEBU[2017-01-12 17:06:23.132102522+01:00] ListImages:
```
the call logs that in ocid logs, and k8s repeats the call some time after but with the same (stuck) result. Notice that when calling `ListImages` with `ocic` you get:
```
DEBU[2017-01-12 17:06:09.078923152+01:00] ListImages: filter:<image:<image:"" > >
```
and it seems to work..at least it won't get stuck.

I'm running with k8s master (which requires also to re-vendor k8s here in `cri-o`) but re-vendoring doesn't solve anything unfortunately .

runcom commented on server/container_list.go:
```diff
@@ -28,6 +28,7 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
 
 // ListContainers lists all containers by filters.
 func (s *Server) ListContainers(ctx context.Context, req *pb.ListContainersRequest) (*pb.ListContainersResponse, error) {
+	s.Update()
```
could you move each of this below the `logrus.Debugf` call, otherwise, it's impossible to understand which call is currently calling `s.Update`

nalind replied:
@runcom okay, repushed with that changed.

nalind commented:
That "ListImages: " debug message suggests that the API client isn't setting the `Filter` field in its `ListImagesRequest`.

runcom commented:
@nalind it seems `ListImages`, `ListPodSandbox` and `ListContainers` timeout when called from k8s (note that k8s does really fast calls one after the other). Snippet from kubelet log:
```
E0112 17:30:23.317131   17460 remote_runtime.go:228] ListContainers with filter "label_selector:<key:\"io.kubernetes.managed\" value:\"true\" > " from runtime service failed: rpc error: code = 4 desc = context deadline exceeded
E0112 17:30:23.317236   17460 kuberuntime_container.go:304] getKubeletContainers failed: rpc error: code = 4 desc = context deadline exceeded
E0112 17:30:23.317248   17460 kubelet.go:1131] Container garbage collection failed: rpc error: code = 4 desc = context deadline exceeded
I0112 17:30:23.317344   17460 server.go:701] Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"127.0.0.1", UID:"127.0.0.1", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'ContainerGCFailed' rpc error: code = 4 desc = context deadline exceeded
E0112 17:30:23.321132   17460 remote_image.go:58] ListImages with filter "<nil>" from image service failed: rpc error: code = 4 desc = context deadline exceeded
E0112 17:30:23.322148   17460 kuberuntime_image.go:106] ListImages failed: rpc error: code = 4 desc = context deadline exceeded
W0112 17:30:23.323381   17460 image_gc_manager.go:164] [imageGCManager] Failed to monitor images: rpc error: code = 4 desc = context deadline exceeded
E0112 17:30:23.322742   17460 remote_image.go:58] ListImages with filter "<nil>" from image service failed: rpc error: code = 4 desc = context deadline exceeded
E0112 17:30:23.323482   17460 kuberuntime_image.go:106] ListImages failed: rpc error: code = 4 desc = context deadline exceeded
```
This is what we have in ocid log:
```
[ omitted ]
DEBU[2017-01-12 17:28:02.829617463+01:00] devmapper: deactivateDevice END()
DEBU[2017-01-12 17:28:02.837024792+01:00] sandboxes: map[]
DEBU[2017-01-12 17:28:02.837058315+01:00] containers: &{map[] {{0 0} 0 0 0 0}}
DEBU[2017-01-12 17:28:23.319179651+01:00] ListContainersRequest filter:<label_selector:<key:"io.kubernetes.managed" value:"true" > >
DEBU[2017-01-12 17:28:23.319330072+01:00] updating sandbox and container information
DEBU[2017-01-12 17:28:23.321837658+01:00] ListImages:
```
Can it be a deadlock somewhere since the calls coming from k8s are really fast? not sure what has introduced this behavior though :/ cause I was able to use k8s just fine ~1month ago.

runcom commented:
> That "ListImages: " debug message suggests that the API client isn't setting the Filter field in its ListImagesRequest.

yeah, right, but why should ocid/c/storage just wait?

nalind commented:
Whoops, that looks like yesterday's change to Update() breaking it.  Reverting that.

nalind commented:
Re-pushed with the changes that made Server.update() block itself removed.

runcom commented:
Unfortunately this is still failing when trying to run an nginx pod, I got two kind of errors:
```
DEBU[2017-01-12 22:43:43.695579003+01:00] devmapper: AddDevice(hash=8369603abb954aa8c9076570cc7853aa0e5967d52f3f1e980740c7c5970f1d8f basehash=)
DEBU[2017-01-12 22:43:43.695614405+01:00] devmapper: Failed to read /var/lib/containers/devicemapper/metadata/8369603abb954aa8c9076570cc7853aa0e5967d52f3f1e980740c7c5970f1d8f with err: open /var/lib/containers/devicemapper/metadata/8369603abb954aa8c9076570cc7853aa0e5967d52f3f1e980740c7c5970f1d8f: no such file or directory
```
and just after running the nginx pod (I had to pull the pause image with ocic before restarting ocid):
```
DEBU[2017-01-12 22:41:36.270567467+01:00] parsed reference into "[devicemapper@/var/lib/containers]docker.io/kubernetes/pause:latest"
2017/01/12 22:41:36 grpc: Server failed to encode response proto: Marshal called with nil
```
In this last case we're probably doing something where the response returned to grpc is nil (I didn't look in the code though)

runcom commented:
The second error from my previous comment is easily (100% for me) reproducible by just trying to start a pod (nginx in my case).

jawnsy commented:
@runcom Do we have heavier "conformance" type tests that would check common apps like nginx, like you're doing? Just wondering if there's some automation we could add to help with that, even if they're only run by CI or only run occasionally (due to being too heavy for normal day-to-day use)

nalind commented:
@runcom We were mistakenly clearing `err` at `sandbox_run.go:74` and `sandbox_run.go:214`, and returning both a nil result and no error isn't something the rpc layer expects.  That should be fixed now.

runcom commented:
> @runcom We were mistakenly clearing err at sandbox_run.go:74 and sandbox_run.go:214, and returning both a nil result and no error isn't something the rpc layer expects. That should be fixed now.

awesome, @sameo was also seeing issues while integrating this PR to work with k8s

nalind commented:
A CRI client can't be expected to know that it needs to tell us to pull the `pause` image, so changed the patch that disables implicit image pulling to still pull `pause` if we need it.

nalind commented:
Rebased.

nalind commented:
Added detail to the error message that we return when we don't have a given image, and didn't pull it.

jjlakis commented:
My image pull problem on kubernetes cluster:
```
  FirstSeen	LastSeen	Count	From			SubObjectPath		Type		Reason		Message
  ---------	--------	-----	----			-------------		--------	------		-------
  18m		18m		1	{default-scheduler }				Normal		Scheduled	Successfully assigned nginx-1978115907-qqxfg to gklab-97a-012
  18m		18m		1	{kubelet gklab-97a-012}				Normal		SandboxReceived	Pod sandbox received, it will be created.
  18m		2m		6	{kubelet gklab-97a-012}	spec.containers{nginx}	Normal		Pulling		pulling image "nginx"
  16m		17s		6	{kubelet gklab-97a-012}	spec.containers{nginx}	Warning		Failed		Failed to pull image "nginx": rpc error: code = 4 desc = context deadline exceeded
  16m		17s		6	{kubelet gklab-97a-012}				Warning		FailedSync	Error syncing pod, skipping: failed to "StartContainer" for "nginx" with ErrImagePull: "rpc error: code = 4 desc = context deadline exceeded"

  16m	4s	22	{kubelet gklab-97a-012}	spec.containers{nginx}	Normal	BackOff		Back-off pulling image "nginx"
  16m	4s	22	{kubelet gklab-97a-012}				Warning	FailedSync	Error syncing pod, skipping: failed to "StartContainer" for "nginx" with ImagePullBackOff: "Back-off pulling image \"nginx\""
```

runcom commented:
@jjlakis could you provide logs from ocid _and_ the kubelet?

runcom commented:
For some reason you're getting a timeout while pulling the image :confused: could a race somewhere?

runcom commented:
mmm this branch isn't working with kubernetes master anymore. I'll update k8s API in master and provide a fix for this PR once @nalind rebase.

runcom commented:
@nalind once https://github.com/kubernetes-incubator/cri-o/pull/321 is merged and you rebase here, could you cherry-pick https://github.com/runcom/ocid-1/commit/080949a842325ace76251eb2aada3b0d3465cce6 to fix containers creation?

runcom commented:
So, with latest k8s master+this PR I'm getting:
```
Events:
  FirstSeen     LastSeen        Count   From                    SubObjectPath           Type            Reason          Message
  ---------     --------        -----   ----                    -------------           --------        ------          -------
  5m            5m              1       {default-scheduler }                            Normal          Scheduled       Successfully assigned nginx-2932712833-czxrb to 127.0.0.1
  5m            5m              1       {kubelet 127.0.0.1}                             Normal          SandboxReceived Pod sandbox received, it will be created.
  3m            3m              1       {kubelet 127.0.0.1}     spec.containers{nginx}  Warning         Failed          Failed to pull image "nginx": rpc error: code = 4 desc = context deadline exceeded
  3m            3m              1       {kubelet 127.0.0.1}                             Warning         FailedSync      Error syncing pod, skipping: failed to "StartContainer" for "nginx" with ErrImagePull: "rpc error: code = 4 desc = context deadline exceeded"

  5m    2m      2       {kubelet 127.0.0.1}     spec.containers{nginx}  Normal  Pulling pulling image "nginx"
  53s   53s     1       {kubelet 127.0.0.1}     spec.containers{nginx}  Normal  Pulled  Successfully pulled image "nginx"
  53s   53s     1       {kubelet 127.0.0.1}     spec.containers{nginx}  Normal  Created Created container with id 2b0d2e527f26207ab1e9ec7c88ff1604cc54cfa5a8744a4d293f115042972867
  53s   53s     1       {kubelet 127.0.0.1}     spec.containers{nginx}  Normal  Started Started container with id 2b0d2e527f26207ab1e9ec7c88ff1604cc54cfa5a8744a4d293f115042972867
```
which means the container came up **_eventually_** (and it works fine!). Still getting that weird `context deadline exceeded` though which is likely the cause of @jjlakis's `ImagePullBackOff`. I'll investigate it further.

runcom commented:
@mrunalp @nalind @sameo @jjlakis figured out the `ImagePullBackOff` issue. It's just a matter of starting the kubelet with `--runtime-request-timeout=5m` (or whatever timeout duration you deem appropriate).

sameo commented:
@runcom I saw my first k8s pod on top of CRI-O with the timeout addition :-D

CRI-O master + this PR, k8s@dca0b2a73d6daf89924c1aa7469bd3ea63cf1f5f + kubelete started with `ca0b2a73d6daf89924c1aa7469bd3ea63cf1f5f`

nalind commented:
Rebased.

jjlakis commented:
Ok, idle cluster (1 master, 2 nodes, one of them with cri-o, v1.5.1) works for:
* this PR with my networking fix #316 
* @runcom's fix-189 branch with newest k8s + my networking fix

by "works" I mean that all pod microservices (kube-proxy, flannel, etc.) are running together with its infra-containers what I can check by `runc list`.

I added `--runtime-request-timeout=5m` to kubelet and `ImagePullBackOff` does not appear, but there's another problem...  Image is pulled, container is created, started, then it crashes and... image is pulled again.
```
  FirstSeen	LastSeen	Count	From			SubObjectPath			Type		Reason		Message
  ---------	--------	-----	----			-------------			--------	------		-------
  3m		3m		1	{default-scheduler }					Normal		Scheduled	Successfully assigned nginxtest-3772885699-cbcnb to gklab-97a-012
  3m		3m		1	{kubelet gklab-97a-012}					Normal		SandboxReceived	Pod sandbox received, it will be created.
  21s		21s		1	{kubelet gklab-97a-012}	spec.containers{nginxtest}	Normal		Pulled		Successfully pulled image "nginx"
  21s		21s		1	{kubelet gklab-97a-012}	spec.containers{nginxtest}	Normal		Created		Created container with id 0d534d33772b573d2c5ff655b9057d1f4ac58c7f32ccf212836efbec9bb9990f
  21s		21s		1	{kubelet gklab-97a-012}	spec.containers{nginxtest}	Normal		Started		Started container with id 0d534d33772b573d2c5ff655b9057d1f4ac58c7f32ccf212836efbec9bb9990f
  3m		20s		2	{kubelet gklab-97a-012}	spec.containers{nginxtest}	Normal		Pulling		pulling image "nginx"
```

Crashed container the node (runc list):
```
default-nginxtest-3772885699-cbcnb-0-infra                         122515      running     /run/containers/aufs-containers/ccb203f2ff14df363259529e7ce83f6d32eb02fe7f40ea17bbe9ec52d81d7247/userdata   2017-01-17T15:08:33.31557929Z
default-nginxtest-3772885699-cbcnb-0-nginxtest-0                   0           stopped     /run/containers/aufs-containers/0d534d33772b573d2c5ff655b9057d1f4ac58c7f32ccf212836efbec9bb9990f/userdata   2017-01-17T15:12:08.282083761Z
```
Assuming that I did something wrong and container crashed because of trivial reason and kubelet simply tries to run it again - why image is pulled again? And what is pretty weird - kubelet decides to send `PullImage` to runtime again (maybe kubelet doesn't know that image is already pulled?). Working on this right now

runcom commented:
> Assuming that I did something wrong and container crashed because of trivial reason and kubelet simply tries to run it again - why image is pulled again? And what is pretty weird - kubelet decides to send PullImage to runtime again (maybe kubelet doesn't know that image is already pulled?). Working on this right now

very likely something wrong with `ImageStatus` reporting the image doesn't exist and k8s tries to pull it again.

nalind commented:
@jjlakis after it crashes, is `ocid` logging another `PullImageRequest`?  If it isn't, then I'm not sure why it's being re-pulled.

jjlakis commented:
@runcom Got the same feeling, adding debug logs to `ImageStatusResponse` right now
@nalind Yes, as I wrote before, kubelet sends `PullImageRequest`

mrunalp commented:
Did you'll check the image pull policy?


> On Jan 17, 2017, at 7:22 AM, JJ Łakis <notifications@github.com> wrote:
> 
> @runcom Got the same feeling, adding debug logs to ImageStatusResponse right now
> @nalind Yes, as I wrote before, kubelet sends PullImageRequest
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

nalind commented:
@runcom cherry-picked your patch, thanks!

mrunalp commented:
I saw this error while testing with k8s:
```
Events:
  FirstSeen     LastSeen        Count   From                    SubObjectPath   Type            Reason          Message
  ---------     --------        -----   ----                    -------------   --------        ------          -------
  13s           13s             1       {default-scheduler }                    Normal          Scheduled       Successfully assigned mynginxtest-2-1578605517-q1bls to 127.0.0.1
  13s           13s             1       {kubelet 127.0.0.1}                     Normal          SandboxReceived Pod sandbox received, it will be created.
  10s           10s             1       {kubelet 127.0.0.1}                     Warning         FailedSync      Error syncing pod, skipping: failed to "CreatePodSandbox" for "mynginxtest-2-1578605517-q1bls_default(048ae8e3-dd12-11e6-acc0-52540003a09a)" with CreatePodSandboxError: "CreatePodSandbox for pod \"mynginxtest-2-1578605517-q1bls_default(048ae8e3-dd12-11e6-acc0-52540003a09a)\" failed: rpc error: code = 2 desc = error creating pod sandbox with name \"default-mynginxtest-2-1578605517-q1bls-0\": Error writing blob: devmapper: Error saving transaction metadata: devmapper: Error creating metadata file: open /var/lib/ocid/devicemapper/metadata/.tmp827313610: no such file or directory"
```

mrunalp commented:
Worked for me after I cleaned up /var/lib/ocid and restarted everything. Needs rebase :)

jawnsy commented:
@nalind _Sorry_ about all the rebase hell :cry: One of my changes renamed `vendor/src` to `vendor`, so if you run into trouble, it may be easiest to remove the commit containing the new dependencies, and re-create them with the new build system. Since it's my fault, I'd be happy to do the rebasing for you if you'd like, and otherwise I'm here if you have questions about what I did in #320

jjlakis commented on server/image_list.go:
```diff
@@ -9,8 +9,25 @@ import (
+			Size_:    result.Size,
+		})
+	}
+	return &response, nil
```
It's good to remain the convention a print debug log with response.

nalind replied:
Agreed.  Fixing.

jjlakis commented on server/image_pull.go:
```diff
@@ -1,86 +1,28 @@
-	// TODO: what else do we need here? (Signatures when the story isn't just pulling from docker://)
-
-	return &pb.PullImageResponse{}, nil
+	return &pb.PullImageResponse{
```
It's good to remain the convention a print debug log with response.

nalind replied:
Agreed.  Fixing.

jjlakis commented on server/image_remove.go:
```diff
@@ -9,5 +11,17 @@ import (
+	if err != nil {
+		return nil, err
+	}
 	return &pb.RemoveImageResponse{}, nil
```
It's good to remain the convention a print debug log with response.

nalind replied:
Agreed, fixing.

jjlakis commented on server/image_status.go:
```diff
@@ -1,16 +1,37 @@
+		}
+		return nil, err
+	}
+	return &pb.ImageStatusResponse{
```
It's good to remain the convention a print debug log with response.

nalind replied:
Agreed, fixing.

nalind commented:
Rebased, adding debug statements in the handlers for image-related requests that log the results.  Updated vendoring results now also pull in `github.com/pkg/errors` and `github.com/opencontainers/go-digest`, and the updated `github.com/opencontainers/runtime-tools` generator's `AddProcessEnv` method now takes two arguments.

nalind commented:
Hmm, I'm not that familiar with travis integration, but it doesn't seem to have run for this last rebase.  https://www.traviscistatus.com/ says everything should be working, though.  Any advice?

runcom commented:
> Hmm, I'm not that familiar with travis integration, but it doesn't seem to have run for this last rebase. https://www.traviscistatus.com/ says everything should be working, though. Any advice?

mmm https://travis-ci.org/kubernetes-incubator/cri-o/pull_requests seems like anything has been triggered from this PR. Could you amend the last commit and re-push? If that doesn't work, it should be something on their end definitely.

nalind commented:
Sure.

runcom commented:
@nalind CI triggered, you can remove that test commit now.

nalind commented:
Okay, CI seems to be back.

mrunalp commented:
LGTM

sameo commented:
LGTM

runcom commented:
LGTM

I'm merging this one right now - there are still issues around devicemapper though and maybe c/storage, we'll tackle these in follow up since I don't believe there's anything left to do as part of this PR (it's better to iterate on simple changes from now on) - for instance, I had to increase the timeout to 15m to be able to run  2 pods spawned at the same time, there seems to be some kind of weird locking which prevents even k8s itself to _ping_ the CRI-O, as said, let's tackle this kind of stuff in follow up. Thanks a lot @nalind and everyone who helped with this :)

jawnsy commented:
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: 

Awesome! Great to see this, looking forward to giving it a spin :smile:
